### ISSUE TITLE
Top Toolbar improvements

### ISSUE THREAD

**Issue by mtias** [WordPress/gutenberg#40450]
The top toolbar has stagnated a bit while the feature set of the editor has evolved. The following design addresses the most pertinent issues including:

1. The lack of parent selector for nested blocks
2. The overall increase in editor UI footprint

<img width="632" alt="Screenshot 2023-01-04 at 11 35 42" src="https://user-images.githubusercontent.com/846565/210546838-3ea8ba01-6391-4024-8421-cc8e9fd3cd03.png">

<details><summary>Original issue</summary>
In particular, the now established block parent selector is missing when top toolbar is enabled, which makes selecting some wrappers particularly challenging (group and gallery, for example). The layout also doesn't feel quite right with the block type not being aligned with the [+], and it becoming somewhat inconvenient on larger screens.

I propose we explore these two changes:

- [ ] Add direct parent selector on the left.
- [ ] Try center layout and/or floating docked toolbar.

If the docked-floated treatment works, we should have to do anything else for the parent selector. If it doesn't, we'd need some visual treatment between the parent and current block types.</summary>

**Comment by jasmussen** [comment]
Three quick options:
<img width="1440" alt="top center" src="https://user-images.githubusercontent.com/1204802/164218077-69a6e7ea-5b38-41a4-ba1a-f26f95cabc94.png">

<img width="1440" alt="top center alt" src="https://user-images.githubusercontent.com/1204802/164218088-9c7a4d52-d564-4690-91f2-881699b66c9b.png">

<img width="1440" alt="top center alt 2" src="https://user-images.githubusercontent.com/1204802/164218092-ff50c08f-2e73-4d25-951d-f0dfac10323f.png">

These ideas are mainly based off inital design outlines suggested in #20592, but with an eye towards toolbar-reuse. In the cases above, the main difference in suggestion 1 is the omission of the drag handle, and in images 2 and 3, the color/radius/padding of the toolbar itself.

**Comment by marceloaof** [comment]
Related: 2.4.1.1 and 2.6.1 https://github.com/WordPress/gutenberg/discussions/36331

**Comment by jameskoster** [comment]
Aligning with the [+] might be safer than aligning to the content due to its variable width. Without that geographical relationship, the floating/docked iteration might not work so well. Option 2 looks quite good to me.

**Comment by pablohoneyhoney** [comment]
Not sure we are there yet. The first could make sense as a natural relationship to the floating mode. The others have a strange alignment. It could be tested centered, for example, or even all the way to the left.

The uncompleted dividers are perhaps an anomaly from the rest of similar cases in the current language.

**Comment by jameskoster** [comment]
In terms of the layout, as we explore concepts here it would be good to consider other edit views like focus mode (https://github.com/WordPress/gutenberg/issues/37946) and the upcoming zoomed-out view (https://github.com/WordPress/gutenberg/pull/41156).

**Comment by david-gonzalez-a8c** [comment]
Hey folks!

Working on a few concepts for this, will share them shortly :)

**Comment by david-gonzalez-a8c** [comment]
Hi folks.

Worked on a couple of prototypes, I'm sharing some videos. I think it's important to see this in movement for the fixed effect and because of the change in sizes of the top bar for the child/parent block, as this could bring some visual awkwardness.

Fixed top bar center-aligned:

https://user-images.githubusercontent.com/92024441/172657942-2dcabc91-5e2f-4b4a-80a6-c34e745774b2.mov

Fixed top bar left-aligned:

https://user-images.githubusercontent.com/92024441/172658022-1e1dc4c8-583d-4ec6-8b10-495087a5fca9.mov

For this option I've tried other alignments: 

![image](https://user-images.githubusercontent.com/92024441/172658262-8fbd5261-97bc-42d4-9bba-6bb215929a10.png)

But the former felt more balanced.

Also to consider are context like these (which in my view rule out the option of merging the block toolbar and the main Gutenberg top menu):
![image](https://user-images.githubusercontent.com/92024441/172659247-faeaae52-5071-4dd1-9d2e-8ef060c2dd1f.png)
Here I suggest the top bar if we implement any of these solutions, would "push down" the content to avoid issues like the floating toolbar is causing:
![image](https://user-images.githubusercontent.com/92024441/172659354-66be975b-b4aa-4d5a-bcb8-6d9529e67c84.png).

Would love to hear everyone's thoughts!

**Comment by jameskoster** [comment]
Another thing to consider is the interplay between top toolbar and select mode. This doesn't feel right:

<img width="995" alt="Screenshot 2022-06-09 at 10 11 57" src="https://user-images.githubusercontent.com/846565/172811368-ce18f13b-cd91-40b5-8126-59eb9b649229.png">

**Comment by jameskoster** [comment]
> which in my view rule out the option of merging the block toolbar and the main Gutenberg top menu

I'm curious to hear more of the thinking behind this. Merging seems like the most consistent approach given the array of different views that exist or are in development:

<img width="740" alt="Screenshot 2022-06-09 at 10 54 33" src="https://user-images.githubusercontent.com/846565/172820235-75c0b902-8d45-46e4-bb34-8b01aa7ee2b3.png">

**Comment by mtias** [comment]
@david-gonzalez-a8c thanks for sharing the mockups. I like the first one, it seems worth trying to see how it feels in practice, as it solves the most salient points right now.

**Comment by david-gonzalez-a8c** [comment]
> I'm curious to hear more of the thinking behind this

Hey @jameskoster. Sorry, that was a little vague on my part! I was referring to efforts in this direction, shared back in https://github.com/WordPress/gutenberg/issues/20592. 

![image](https://user-images.githubusercontent.com/92024441/174280358-45a6c052-8761-4d4a-9654-3a79d7ea2195.png)

Not only because of contexts where there's a selector in the main navigation, but also I think it's challenging when it comes to information density.

What you shared makes total sense as a solution to me! I think it's a matter of pros and cons now: I personally like the simplicity of extending the main bar like what you shared and how unambiguous is it, with the floating toolbar it may not be obvious that it will stick to the top until users scroll or play around. But then this is probably more convenient for bigger screens and I think the ability to share the same component in both modes could be valuable.

**Comment by jameskoster** [comment]
When top toolbar is enabled, the document tools could morph into block tools on selection:

![top toolbar](https://user-images.githubusercontent.com/846565/196959004-fa2d897a-833d-4927-8fe3-e4b9d84e1880.gif)

It won't scale down so well to small screens, so we may have to revert to stacking there.

**Comment by mtias** [comment]
@jameskoster I think this could work, but how would it display the parent block? That's one of the main things this needs to address. Also we might need a way to dismiss the selection and get to the list view / undo redo tools from the toolbar itself. What if the [+] becomes a [<], for example? (or some other collapse icon)

**Comment by jameskoster** [comment]
> What if the [+] becomes a [<], for example? (or some other collapse icon)

Then you wouldn't be able to open the Inserter (keyboard aside) when a block is selected, which could be a can of worms?

I find that clicking in empty space to deselect is a good heuristic, both in the canvas and in empty regions of the UI. There's also the `esc` behaviour to fall back on.

As for the parent selector, it's probably just an exercise in finding the right separator. Some of the examples already shared could work, here are a few others:

<img width="551" alt="Screenshot 2022-11-21 at 13 22 07" src="https://user-images.githubusercontent.com/846565/203065801-11dbfe84-03fe-4f7e-b71a-d6b898591d77.png">

<img width="450" src="https://user-images.githubusercontent.com/846565/203065821-019a7ced-bdd2-40ae-8fc6-dcf45b361ea3.gif">

<img width="450" src="https://user-images.githubusercontent.com/846565/203065824-88db527c-183e-47ee-84bf-f5d81348256e.gif">

**Comment by mtias** [comment]
To clarify, the [<] wouldn't unselect a block, just show the higher level tools.

**Comment by jameskoster** [comment]
Yes I suppose that could work:


https://user-images.githubusercontent.com/846565/203281416-1048d3a4-5626-4744-ba67-dc5e376b42e0.mp4

There' a question about how you get _back_ to the block toolbar, but clicking the block again seems fairly intuitive.

**Comment by jasmussen** [comment]
A couple of thoughts for when we're able to pick this back up:

* Latest iteration by looks close in terms of behavior:
  * Back button shows the inserter and other tools again
  * Mini-breadcrumbs for selecting parent block
  * A single unified toolbar
* But it needs probably needs to show the inserter always. 
  * This is to consider the flow, where you might click a paragraph inside a group, use breadcrumbs to select the group, and then open the inserter to insert a block there. 
  * So we'd show Inserter | Back | Breadcrumbs/Block type | Tools.
  * It doesn’t deselect.
  * The back button could be more visible or obvious, the light gray isn’t clear enough. The solution might be a different icon, the "back" arrow suggests history, but in fact it's a "show more tools" button — simply showing again the undo/redo/list view buttons that were covered by the block toolbar.
* The parent/child relationship could be a bit clearer.
  * Perhaps the separator needs to not be an arrow, maybe a dot or a smaller chevron, and perhaps not light gray. 
  * Specifically it could be clearer what semantically is the parent: "this is the anchor, and this is the child". Maybe the parent has the same footprint as the inserter plus, but is dark?
  * Or maybe it has the light gray background, like the current back button has?
* Just like how this top toolbar replaces the shortcuts in the top left when a block is selected, there is a past idea for the inserter to similarly show some commonly used blocks that space. The overlap here is worth keeping in mind: does this toolbar mechanism replace or augment the other idea?

It would be good to bring home this design so we can get it developed. @jameskoster do you have a Figma handy? Would be happy to take a stab at the above.

**Comment by jameskoster** [comment]
It's a bit messy, but here's the [Figma](https://www.figma.com/file/eGoqmWIztNBcib1ukrFwki/Explorations?node-id=1355%3A48662&t=cmIPpNq1e8SdCjN0-1).

**Comment by jasmussen** [comment]
[Took a stab](https://www.figma.com/file/vmHxwa8WhiyiP9poe2om9J/Inspector-Panels?node-id=13916%3A90885&t=T03YEkwwGYQedGoV-1) at parsing the feedback above. At face value, it falls apart:

<img width="1384" alt="eh" src="https://user-images.githubusercontent.com/1204802/206430749-6fe81d4b-8ac4-4ee9-8263-1f3df3bc817b.png">

The main issue above is the double close icons in the inserter open + block selected case, as well as the fact that the more buttons have a background, the more clumsy it gets. But the dark color or at least some "anchoring background" on the parent block still feels important to portray the relationship.

So I tried a variation of that which took it in a different direction:

<img width="1384" alt="maybe" src="https://user-images.githubusercontent.com/1204802/206430958-56e73c24-c4fd-4caf-a553-95c912dccc6f.png">

Maybe?

**Comment by jasmussen** [comment]
Just a quick followup after a slack huddle with Jay, some options on the parent/child unit that can ideally serve as an anchoring element:
<img width="660" alt="Frame 18" src="https://user-images.githubusercontent.com/1204802/206450794-b01db866-83af-4411-b768-7e7340352ed6.png">

I'm not sure either of these entirely work, but there's something about 3 that could be interesting. Sharing in case it inspires ideas.

**Comment by mtias** [comment]
Can we look at it without the dark toggled state for now? That seems like a separate thing to consider.

**Comment by jasmussen** [comment]
Perhaps this, then?

<img width="1384" alt="maybe, i2" src="https://user-images.githubusercontent.com/1204802/206669050-c5ecb1bc-df54-4ea7-8b49-022d9f535881.png">

**Comment by jameskoster** [comment]
Here's a version with some animation. I wanted to see if it would help communicate the relationship between document + block tools, and help define the up-arrow icon.

![animate](https://user-images.githubusercontent.com/846565/206693281-fbb9458e-ce44-44a9-b286-43e7d6eda742.gif)

It's a bit tenuous when it comes to the Inspector interactions, but I figured it was worth sharing anyway.

**Comment by pablohoneyhoney** [comment]
Looking promising to me. 

Though I think the shortcuts are adding redundancy and confusion in that context that already serves for two different mental models (full canvas controls + block controls), and it'd be adding yet a third (block addition or shortcuts). Too many elements moving too. 

Either we move the shortcuts out of that top bar context and place them below the trigger (+) or they have a very different visual footprint so they actually have a different intent than the other two models.  

<img width="221" alt="Screen Shot 2022-12-13 at 11 09 26 AM" src="https://user-images.githubusercontent.com/25516611/207385021-60a8c1e3-2865-43f1-b533-83d469533a76.png">

**Comment by jasmussen** [comment]
For me the jury's out on whether showing recently used blocks when the inserter is open is actually useful. I think it's good to have a design ready that considers it, but it also seems smart to potentially start without it, like so:

<img width="1384" alt="maybe, i3" src="https://user-images.githubusercontent.com/1204802/207546805-0fc568ea-dbfe-4afa-9d3b-a9b5dac377d4.png">

Then we'll know the design scales if we choose to extend it in that direction. And if not, that works too.

**Comment by mtias** [comment]
The shortcuts is not strictly related tot his issue either, it's something we should do and explore separately, just made sense to consider it in the design explorations. Can we narrow down the design proposal to the most basic needs? I think we are close, but some of the visual treatments (like wrapping the block + parent) seem a bit arbitrary.

**Comment by jasmussen** [comment]
Here's a variant that uses vertical separators:

<img width="1384" alt="maybe, i4" src="https://user-images.githubusercontent.com/1204802/208076707-5a8d2880-73ae-4cad-910f-c9d3311eec8b.png">

**Comment by richtabor** [comment]
A couple thoughts: 

1. Does this proposal require de-selecting & re-selecting the block to re-render the block tools?
2. I'm hesitant to bury the List view control, once a block is selected. It'd require you to either press the "up" icon, then the list view icon to render — or de-select the current block. Whereas today, List view can be toggled without breaking the edit flow. List view does not persist either, which compounds this.

**Comment by jasmussen** [comment]
> Does this proposal require de-selecting & re-selecting the block to re-render the block tools?

The little "left/up" arrow would take you back to the "top level" tools.

> I'm hesitant to bury the List view control, once a block is selected. It'd require you to either press the "up" icon, then the list view icon to render — or de-select the current block. Whereas today, List view can be toggled without breaking the edit flow. List view does not persist either, which compounds this.

For my personal workflow the list view is something I toggle once and leave it there. I can see that it gets untoggled when you open the inserter, but I'm not convinced it's a problem with the extra click — you do save a great deal of space, after all.

That said, I think it should be easier to deselect. I think escape for select mode might be a keyboard shortcut to revisit. It could also be interesting to see if there was a way to make the inserter and list view have more in common, since they are both toggles on the left.

**Comment by richtabor** [comment]
> The little "left/up" arrow would take you back to the "top level" tools.

Right. But per the visual above, the “left/up” arrow goes away once clicked - so there’s no way to return to block tools, unless you click off of the block, and back on it. Yea?

**Comment by jasmussen** [comment]
Right, you'd have to click the block again.

**Comment by mtias** [comment]
Since we don't want that button to mean "unselect", it might be worth exploring the opposite icon at the end of the default set of header tools to pull back in the block toolbar.

**Comment by jasmussen** [comment]
> it might be worth exploring the opposite icon at the end of the default set of header tools to pull back in the block toolbar.

Trying this out in a very literal sense:

<img width="772" alt="i5 swapping between parent and block tools" src="https://user-images.githubusercontent.com/1204802/208858396-6613ae4b-8e80-4b6a-9cf1-9cce6fdfa272.png">

**Comment by mtias** [comment]
Maybe the currently selected block icon is more clear / better.

**Comment by jasmussen** [comment]
I can see it working, especially if we can add some nice expanding/collapsing reveal animation:

<img width="772" alt="i5 swapping between parent and block tools" src="https://user-images.githubusercontent.com/1204802/209093876-08fc6c32-8395-4fde-a431-0a742653ce4b.png">

**Comment by mtias** [comment]
Yes, I think this is worth trying. It could also be the block icon plus a small arrow, or the small triangle indicator we had a while ago, but we can refine if the mechanics work well.

**Comment by tellthemachines** [comment]
Added the "needs a11y feedback" label to make sure we consider all aspects of the desired functionality. For example:

* What happens to the list view button if list view is open when the block is selected? (we may actually be using list view to select the block)
* What relationship does the up arrow have with global tools? E.g. should global tools be a popover-type component? Likewise with down arrow and the block toolbar.

**Comment by tellthemachines** [comment]
Some quick mockups achieved by moving stuff around in dev tools, so they don't include buttons to switch between toolbars yet, but trying to get a feel for what it'll look like when the block toolbar as a lot of controls. This is the button block toolbar:

In the post editor, with icons and with text labels:
<img width="906" alt="Screenshot 2023-01-19 at 3 27 27 pm" src="https://user-images.githubusercontent.com/8096000/213382782-47e6e18f-df5b-4eff-93d0-c6d4b565d247.png">
<img width="907" alt="Screenshot 2023-01-19 at 3 28 41 pm" src="https://user-images.githubusercontent.com/8096000/213382810-4a834e2a-a808-4c5e-9d5c-8728c629628a.png">

In the site editor, with icons and with text labels:



<img width="981" alt="Screenshot 2023-01-19 at 3 22 43 pm" src="https://user-images.githubusercontent.com/8096000/213382882-caecacf8-294e-4ac6-bd65-b6ead8916dae.png">

<img width="992" alt="Screenshot 2023-01-19 at 3 31 01 pm" src="https://user-images.githubusercontent.com/8096000/213382900-aa8899fa-ee09-41db-9065-7694ef2e6ac1.png">

The text labels in particular need more space, so what can we do with big toolbars, or at small breakpoints where there isn't much space? Should we hide some of the block tools Thoughts appreciated!

**Comment by draganescu** [comment]
One nice thing about putting the block tools to the top bar is that it gives us block tools in distraction free mode:

https://user-images.githubusercontent.com/107534/213403102-f04c5d88-e982-4cea-8eb6-e2bdc5565616.mp4

**Comment by draganescu** [comment]
A problem that needs to be addressed in the design @jasmussen is that, as far as a quick hacky implementation shows, if you have this implemented, using the up arrow have shown the document tools, then opened inserter or list view, then went back to block tools with the button on the right, then you won’t see the block tools dropdowns:

<img width="1342" alt="Screenshot 2023-01-19 at 17 46 59" src="https://user-images.githubusercontent.com/107534/213410681-7fe967d0-3fac-405c-be0d-8c9481dd5297.png">

The block switcher is open behind the list view in the screenshot above. Should they just open on top? If so we may have a z-index war with the interface package's panels.

**Comment by jasmussen** [comment]
> using the up arrow have shown the document tools, then opened inserter or list view, then went back to block tools with the button on the right, then you won’t see the block tools dropdowns:

Shouldn't the block switcher should be at a higher level? I see that dropdown as no different than the dropdown that extends from the Preview button.

As far as text labels go, perhaps it's time we create a tracking issue. Spread as feedback across a number of tickets, it's insufficient to just convert aria text to a button text label for this, we need to carefully think about the titles of each of those controls as a starting point.

**Comment by aaronrobertshaw** [comment]
I've created a tracking issue (https://github.com/WordPress/gutenberg/issues/47723) to help coordinate work related to achieving the goals and design outlined here.

**Comment by draganescu** [comment]
@jasmussen 
What happens on small viewports? Nothing or something new?


https://user-images.githubusercontent.com/107534/216611252-20503f31-da40-44a7-ba16-75ce0970eca3.mp4

**Comment by jasmussen** [comment]
Yes, that's a good question. To help the discussion I'll refer to the top toolbar as two ingredients: document and block controls, which for the purposes of this issue are merged into one on desktop. My instinct for the time being is that on mobile, we separate those two and stack them, block controls below document controls.

I think in general, the mobile toolbar is in due of separate improvements. But even with tweaks there, it's unlikely we'll ever fit block controls and document controls on a single row. In that light, another improvement that's worth revisiting is to see if we can affix the block controls to the top of the mobile device soft-keyboard. Native apps do this very well, but so far this has been very tricky to do in mobile web-apps outside of Chrome on Android. To my understanding, that's due to Android resizing the _viewport_ to fit the available space remaining after the soft keyboard opens, meaning we can just fix the toolbar to the bottom on such touch devices. Meanwhile in some other browsers, like Safari on iOS, the viewport size does not change (last I checked), and instead the soft-keyboard _covers_ part of the page. It's worth looking into this behavior in some codepens, because I believe some of the opportunities here have changed in recent years, it's at least 2 years since I looked at this the last time. 

By the way, when we do get to that rabbit hole, this trick might help:

```
@media (hover: none) {
	...
}
```
